sklearn/ensemble/_gb.py

```diff
@@ -23,33 +23,246 @@
+            elif isinstance(loss, HalfMultinomialLoss):
+                # we take advantage that: y - prob = residual
+                residual_ = residual.take(indices, axis=0)
+                prob = y_ - residual_
+                K = loss.n_classes
+                # numerator = negative gradient * (k - 1) / k
+                # Note: The factor (k - 1)/k appears in the original papers "Greedy
+                # Function Approximation" by Friedman and "Additive Logistic
+                # Regression" by Friedman, Hastie, Tibshirani. This factor is, however,
+                # wrong or at least arbitrary as it directly multiplies the
+                # learning_rate. We keep it for backward compatibility.
+                numerator = np.average(residual_, weights=sw)
+                numerator *= (K - 1) / K
+                # denominator = (diagonal) hessian = prob * (1 - prob)
+                denominator = np.average(prob * (1 - prob), weights=sw)
+
+                # prevents overflow and division by zero
+                sw_sum = indices.shape[0] if sw is None else np.sum(sw)
+                if abs(denominator * sw_sum) < 1e-150:
+                    tree.value[leaf, 0, 0] = 0.0
+                else:
+                    tree.value[leaf, 0, 0] = numerator / denominator
+            elif isinstance(loss, ExponentialLoss):
+                z = 2.0 * y_ - 1.0  # z is -1 or +1
+                raw_pred = raw_prediction[indices, k]
+                # numerator = negative gradient
+                numerator = np.average(z * np.exp(-z * raw_pred), weights=sw)
+                # denominator = hessian
+                denominator = np.average(np.exp(-z * raw_pred), weights=sw)
+
+                # prevents overflow and division by zero
+                sw_sum = indices.shape[0] if sw is None else np.sum(sw)
+                if abs(denominator * sw_sum) < 1e-150:
+                    tree.value[leaf, 0, 0] = 0.0
+                else:
+                    tree.value[leaf, 0, 0] = numerator / denominator
+            else:
+                update = loss.fit_intercept_only(
+                    y_true=y_ - raw_prediction[indices, k],
+                    sample_weight=sw,
+                )
+                tree.value[leaf, 0, 0] = update
+
+    # update predictions (both in-bag and out-of-bag)
+    raw_prediction[:, k] += learning_rate * tree.value[:, 0, 0].take(
+        terminal_regions, axis=0
+    )
+
+
+def set_huber_delta(loss, y_true, raw_prediction, sample_weight=None):
+    """Calculate and set self.closs.delta based on self.quantile."""
+    abserr = np.abs(y_true - raw_prediction.squeeze())
+    if sample_weight is None:
+        # Same as np.quantile(abserr, loss.quantile, axis=0, method="inverted_cdf")
+        delta = percentile(abserr, 100 * loss.quantile, axis=0, method="lower")
+    else:
+        delta = _weighted_percentile(abserr, sample_weight, 100 * loss.quantile)
```

Comment:
Very strange to not have coverage here. If you have a chance, can you open a new PR to add a test to check huber loss & sample_weights?

Reply:
I opened the issue #26960.